### PR TITLE
HDDS-5701. Add auditLog when DN register fail

### PR DESCRIPTION
**Pull request by wzhallright**
## What changes were proposed in this pull request?

I recently upgraded the scm, after upgrade I found the dn can not register to scm.
So, I checked the audit log and want to know why it failed，the audit log like this
`2021-08-31 14:47:07,920 | INFO  | SCMAudit | user=ozone | ip=localhost | op=REGISTER {datanodeDetails=161de39c-0ffe-4d39-8f81-f66fbd8148a7{ip: localhost, host: , ports: [REPLICATION=9886, RATIS=9858, RATIS_ADMIN=9857, RATIS_SERVER=9856, STANDALONE=9859], networkLocation: /default-rack, certSerialId: null, persistedOpState: null, persistedOpStateExpiryEpochSec: 0}} | ret=SUCCESS |`
the audit log is still register success, but the real thing is that the dn registration failed.
The register dn layout version lower than scm version, it may be that we need to log the failure in the audit.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5701

### CODE REVIEW

**Comment by ChenSammi** [comment]
@wzhallright ,  what is the log after the patch, would you paste an example here?

**Comment by wzhallright** [comment]
> @wzhallright , what is the log after the patch, could you paste an example here?

Thanks for review.
The fail log is

`2021-09-01 17:41:09,660 | ERROR | SCMAudit | user=ozone | ip=localhost | op=REGISTER {datanodeDetails=d68d781c-e720-4107-8b47-eb59a5e6cfd7{ip: localhost, host: , ports: [REPLICATION=9886, RATIS=9858, RATIS_ADMIN=9857, RATIS_SERVER=9856, STANDALONE=9859], networkLocation: /default-rack, certSerialId: null, persistedOpState: IN_SERVICE, persistedOpStateExpiryEpochSec: 0}} | ret=FAILURE |`

If I modify it like this PR，when getRegisterResponse(registeredCommand) throw exception, the audit will log twice.
So, I'm wondering if I need to add a boolean parameter for ErrorCode.errorNodeNotPermitted when log audit in finally.
Do you have any suggestions?

**Comment by wzhallright** [comment]
@ChenSammi @ayushtkn Could you help review this PR? Thanks.

**Review by ayushtkn — COMMENTED** [review]
Just going through it, A general question, Why don't we throw an exception itself when the result isn't success, and we can cover that check inside the try block, and the audit log will get handled?
Something like :
https://github.com/apache/ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/endpoint/RegisterEndpointTask.java#L165

This exception anyway will get thrown if we don't have a success, we can have this check inside the try block itself?

**Comment by wzhallright** [comment]
@ayushtkn Thank you so much for review, I ignored the dn will throw exception anyway when we don't have a success.
I resubmitted a pr, could you help review again, thank you very much.

**Review by ayushtkn — APPROVED** [review]
LGTM

**Comment by ayushtkn** [comment]
@ChenSammi do you have any further comments?

**Comment by wzhallright** [comment]
> @ChenSammi do you have any further comments?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java [hunk 1]
```diff
@@ -241,6 +241,9 @@ public SCMRegisteredResponseProto register(
                       pipelineReportsProto));
     }
     try {
+      Preconditions.checkState(registeredCommand.getError()
+              == SCMRegisteredResponseProto.ErrorCode.success,
+          "DataNode has higher Software Layout Version than SCM.");
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:246]: While this one-liner is elegant from code perspective, I think we should avoid polluting the audit log with the exception.

**Review by adoroszlai — COMMENTED** [review]
Thanks @wzhallright for the patch.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:246]: @wzhallright Would you mind adding a slightly different check, which does not log an exception unnecessarily?

**Pull request closed without merging**